File: ts/src/session.ts
/**
 * FileSession — record/replay/passthrough session.
 */
import { ShapeMismatchError } from "./stream.js";
import {
  OccurrenceCounter,
  type StreamOpen,
  streamCanonical,
  streamFingerprint,
} from "./streamfp.js";
import {
  type StreamDirection,
  type StreamScrubFn,
  type StreamScrubInfo,
  scrubFrame,
} from "./streamScrub.js";
import {
  type StreamCassette,
  StreamRecording,
  StreamReplay,
} from "./streamSession.js";
import { type Adapter, type Cassette, ErrCassetteMiss, type Mode, type Session } from "./xrr.js";

export class FileSession implements Session {
  /**
   * Per-session occurrence counter for streamed opens — one session object
   * is one counter domain (spec/cassette-format-streaming.md). Streaming
   * adapters key it by their identifying tuple at each stream open.
   */
  readonly streamCounter = new OccurrenceCounter();

  /**
   * Creates a session. `streamScrub` installs the frame-level scrub hook
   * (see streamScrub.ts): every streamed frame passes through it at record
   * time and every live send passes through it at replay time. Install the
   * SAME hook when recording and when replaying — scrubbing is symmetric by
   * design, and a session replaying a scrubbed cassette without the hook
   * fails with a stream mismatch.
   */
  constructor(
    private readonly mode: Mode,
    private readonly cassette: Cassette,
    private readonly streamScrub?: StreamScrubFn
  ) {}

  /**
   * The session mode. Streaming adapters dispatch on it to pick the record,
   * replay, or passthrough path at stream open.
   */
  get sessionMode(): Mode {
    return this.mode;
  }

  /**
   * Applies the session's frame scrub hook to data, returning data
   * unchanged when no hook is installed.
   *
   * Adapters whose open identity derives from message bytes (the gRPC
   * server-stream msg_hash) MUST compute the derived identity over this
   * method's output, in record and replay mode alike, so both modes address
   * the cassette by the scrubbed content. Frames handed to the core
   * (recordSend / recordRecv, replay send) are scrubbed by the core
   * itself — adapters pass them raw and never double-scrub.
   */
  scrubStreamFrame(dir: StreamDirection, info: StreamScrubInfo, data: Uint8Array): Uint8Array {
    return scrubFrame(this.streamScrub, dir, info, data);
  }

  async record<Req, Resp>(
    adapter: Adapter<Req, Resp>,
    req: Req,
    do_: () => Promise<Resp>
  ): Promise<Resp> {
    switch (this.mode) {
      case "record":
        return this.doRecord(adapter, req, do_);
      case "replay":
        return this.doReplay(adapter, req);
      case "passthrough":
        return do_();
      default: {
        const exhaustive: never = this.mode;
        throw new Error(`xrr: unknown mode "${exhaustive}"`);
      }
    }
  }

  private async doRecord<Req, Resp>(
    adapter: Adapter<Req, Resp>,
    req: Req,
    do_: () => Promise<Resp>
  ): Promise<Resp> {
    const resp = await do_();
    const fp = await adapter.fingerprint(req);
    await this.cassette.save(
      adapter.id,
      fp,
      adapter.serializeReq(req),
      adapter.serializeResp(resp)
    );
    return resp;
  }

  private async doReplay<Req, Resp>(
    adapter: Adapter<Req, Resp>,
    req: Req
  ): Promise<Resp> {
    const fp = await adapter.fingerprint(req);
    let loaded: { req: unknown; resp: unknown };
    try {
      loaded = await this.cassette.load(adapter.id, fp);
    } catch (err) {
      if (err === ErrCassetteMiss) throw ErrCassetteMiss;
      throw err;
    }
    return adapter.deserializeResp(loaded.resp);
  }

  // ── streamed interactions ────────────────────────────────────────────────

  /**
   * Opens a streamed interaction for recording. The adapter observes the
   * live stream and mirrors it into the returned recording: recordSend /
   * recordRecv per message, recordHalfClose when the client closes its send
   * side, then finish exactly once when the terminal is observed — only
   * finish persists the pair.
   */
  async openStreamRecord(open: StreamOpen): Promise<StreamRecording> {
    const cassette = this.checkStreamOpen(open, "record", "openStreamRecord");
    const { fp, n } = this.streamOpenFingerprint(open);
    const payload: Record<string, unknown> = { ...open.payload };
    // Informational occurrence ordinal: recoverable from disk, never read
    // back to drive matching.
    if (n >= 0) payload.n = n;
    return new StreamRecording(
      cassette,
      open.adapterID,
      fp,
      open.type,
      payload,
      this.streamScrub
    );
  }

  /**
   * Locates the cassette pair for a streamed open and returns a replay
   * handle. Throws ErrCassetteMiss when no pair exists and
   * ShapeMismatchError when the pair is unary or its recorded stream type
   * differs. The occurrence counter is consumed exactly as in record mode,
   * hit or miss.
   */
  async openStreamReplay(open: StreamOpen): Promise<StreamReplay> {
    const cassette = this.checkStreamOpen(open, "replay", "openStreamReplay");
    const { fp } = this.streamOpenFingerprint(open);
    const pair = await cassette.loadStreamed(open.adapterID, fp);
    if (pair.req.stream.type !== open.type) {
      throw new ShapeMismatchError(
        `xrr: shape mismatch: recorded stream type "${pair.req.stream.type}", requested "${open.type}"`
      );
    }
    return new StreamReplay(fp, pair, this.streamScrub, open.adapterID);
  }

  private checkStreamOpen(open: StreamOpen, want: Mode, verb: string): StreamCassette {
    if (this.mode !== want) {
      throw new Error(`xrr: ${verb} requires ${want} mode (session is "${this.mode}")`);
    }
    if (!open.adapterID) {
      throw new Error(`xrr: ${verb} requires an adapter id`);
    }
    const c = this.cassette as Cassette & Partial<StreamCassette>;
    if (typeof c.saveStreamed !== "function" || typeof c.loadStreamed !== "function") {
      throw new Error(`xrr: ${verb} requires a stream-capable cassette`);
    }
    return c as Cassette & StreamCassette;
  }

  /**
   * Computes the open-time fingerprint, consuming the occurrence counter
   * for counter-addressed opens. The counter is keyed by the adapter id
   * plus the canonical identity (sans "n"), i.e. the adapter's identifying
   * tuple. n is -1 for content-addressed opens.
   */
  private streamOpenFingerprint(open: StreamOpen): { fp: string; n: number } {
    let n = -1;
    if (open.counter) {
      n = this.streamCounter.next(open.adapterID, streamCanonical(open, -1));
    }
    return { fp: streamFingerprint(open, n), n };
  }
}

File: rs/src/session.rs
use crate::{
    cassette::FileCassette,
    error::XrrError,
    stream::StreamCounters,
    stream_scrub::{scrub_frame, StreamDirection, StreamScrub, StreamScrubInfo},
    Adapter,
};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Mode {
    Record,
    Replay,
    Passthrough,
}

pub struct Session {
    mode: Mode,
    cassette: FileCassette,
    stream_counters: StreamCounters,
    stream_scrub: Option<StreamScrub>,
}

impl Session {
    pub fn new(mode: Mode, cassette: FileCassette) -> Self {
        Self {
            mode,
            cassette,
            stream_counters: StreamCounters::new(),
            stream_scrub: None,
        }
    }

    /// A session whose streamed interactions pass every frame through
    /// `scrub`. Identical to [`Session::new`] when the hook is absent:
    /// frames record and replay verbatim.
    ///
    /// Install the SAME hook when recording and when replaying: scrubbing
    /// is symmetric by design (see [`crate::stream_scrub::StreamScrub`]),
    /// and a session replaying a scrubbed cassette without the hook fails
    /// with a stream mismatch.
    pub fn with_stream_scrub(mode: Mode, cassette: FileCassette, scrub: StreamScrub) -> Self {
        Self {
            mode,
            cassette,
            stream_counters: StreamCounters::new(),
            stream_scrub: Some(scrub),
        }
    }

    pub(crate) fn stream_scrub(&self) -> Option<&StreamScrub> {
        self.stream_scrub.as_ref()
    }

    /// Apply the session's frame scrub hook to `data`, returning it
    /// unchanged when no hook is installed.
    ///
    /// Adapters whose open identity derives from message bytes (the gRPC
    /// server-stream `msg_hash`) MUST compute the derived identity over
    /// this function's output, in record and replay mode alike, so both
    /// modes address the cassette by the scrubbed content. Frames handed to
    /// the core (`record_send`/`record_recv`, replay `send`) are scrubbed
    /// by the core itself — adapters pass them raw and never double-scrub.
    pub fn scrub_stream_frame(
        &self,
        dir: StreamDirection,
        info: &StreamScrubInfo,
        data: &[u8],
    ) -> Vec<u8> {
        scrub_frame(self.stream_scrub.as_ref(), dir, info, data)
    }

    pub fn mode(&self) -> Mode {
        self.mode
    }

    pub(crate) fn cassette(&self) -> &FileCassette {
        &self.cassette
    }

    /// Occurrence counters for streamed fingerprints: one session object
    /// is one counter domain, counted identically in record and replay.
    pub fn stream_counters(&self) -> &StreamCounters {
        &self.stream_counters
    }

    pub fn record<A: Adapter>(
        &self,
        adapter: &A,
        req: &A::Req,
        do_: impl FnOnce() -> Result<A::Resp, XrrError>,
    ) -> Result<A::Resp, XrrError> {
        match self.mode {
            Mode::Record => {
                let resp = do_()?;
                let fp = adapter.fingerprint(req)?;
                self.cassette.save(adapter.id(), &fp, req, &resp)?;
                Ok(resp)
            }
            Mode::Replay => {
                let fp = adapter.fingerprint(req)?;
                let (_req, resp): (A::Req, A::Resp) =
                    self.cassette.load(adapter.id(), &fp)?;
                Ok(resp)
            }
            Mode::Passthrough => do_(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::adapters::exec::{ExecAdapter, ExecRequest, ExecResponse};
    use std::collections::HashMap;
    use tempfile::TempDir;

    fn req() -> ExecRequest {
        ExecRequest {
            argv: vec!["echo".into(), "hello".into()],
            stdin: "".into(),
            env: HashMap::new(),
        }
    }

    fn resp() -> ExecResponse {
        ExecResponse {
            stdout: "hello\n".into(),
            stderr: "".into(),
            exit_code: 0,
            duration_ms: 1,
        }
    }

    #[test]
    fn record_saves_and_returns() {
        let tmp = TempDir::new().unwrap();
        let cassette = FileCassette::new(tmp.path());
        let session = Session::new(Mode::Record, cassette);
        let adapter = ExecAdapter;
        let r = req();

        let result = session.record(&adapter, &r, || Ok(resp())).unwrap();
        assert_eq!(result.stdout, "hello\n");

        // Verify file was written.
        let fp = adapter.fingerprint(&r).unwrap();
        let path = tmp
            .path()
            .join(format!("exec-{}.req.yaml", fp));
        assert!(path.exists());
    }

    #[test]
    fn replay_loads_without_calling_do() {
        let tmp = TempDir::new().unwrap();
        let adapter = ExecAdapter;
        let r = req();
        let fp = adapter.fingerprint(&r).unwrap();

        // Pre-save cassette files.
        let cassette = FileCassette::new(tmp.path());
        cassette.save("exec", &fp, &r, &resp()).unwrap();

        let cassette2 = FileCassette::new(tmp.path());
        let session = Session::new(Mode::Replay, cassette2);

        let mut called = false;
        let result = session
            .record(&adapter, &r, || {
                called = true;
                Ok(resp())
            })
            .unwrap();

        assert!(!called, "do_ should not be called in replay mode");
        assert_eq!(result.stdout, "hello\n");
    }

    #[test]
    fn replay_miss_returns_cassette_miss() {
        let tmp = TempDir::new().unwrap();
        let cassette = FileCassette::new(tmp.path());
        let session = Session::new(Mode::Replay, cassette);
        let adapter = ExecAdapter;
        let r = req();

        let result = session.record(&adapter, &r, || Ok(resp()));
        assert!(matches!(result, Err(XrrError::CassetteMiss { .. })));
    }

    #[test]
    fn passthrough_calls_do_without_saving() {
        let tmp = TempDir::new().unwrap();
        let cassette = FileCassette::new(tmp.path());
        let session = Session::new(Mode::Passthrough, cassette);
        let adapter = ExecAdapter;
        let r = req();

        let mut called = false;
        let result = session.record(&adapter, &r, || {
            called = true;
            Ok(resp())
        }).unwrap();

        assert!(called);
        assert_eq!(result.exit_code, 0);

        // No files should exist.
        let entries: Vec<_> = std::fs::read_dir(tmp.path())
            .unwrap()
            .collect();
        assert!(entries.is_empty());
    }
}
